### PR TITLE
fix: incorrect conversion between integer types

### PR DESCRIPTION
**Pull request by tenthirtyam**
### Summary

Fixes incorrect conversion of an unsigned 32-bit integer from  to a lower bit size type int32 without an upper bound check.

### Testing

```console
packer-plugin-vsphere on  fix/incorrect-conversion via 🐹 v1.23.0 
➜ go fmt ./...

packer-plugin-vsphere on  fix/incorrect-conversion via 🐹 v1.23.0 
➜ make test   
?       github.com/hashicorp/packer-plugin-vsphere      [no test files]
?       github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/common/testing [no test files]
ok      github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/clone        2.267s
?       github.com/hashicorp/packer-plugin-vsphere/examples/driver      [no test files]
?       github.com/hashicorp/packer-plugin-vsphere/version      [no test files]
ok      github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/common       4.408s
ok      github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/driver       12.610s
ok      github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/iso  4.657s
ok      github.com/hashicorp/packer-plugin-vsphere/builder/vsphere/supervisor   12.468s
ok      github.com/hashicorp/packer-plugin-vsphere/post-processor/vsphere       6.774s
ok      github.com/hashicorp/packer-plugin-vsphere/post-processor/vsphere-template4.488s
```

### Reference

- [x] https://github.com/hashicorp/packer-plugin-vsphere/security/code-scanning/3
- [x] https://github.com/hashicorp/packer-plugin-vsphere/security/code-scanning/4
- [x] https://github.com/hashicorp/packer-plugin-vsphere/security/code-scanning/5

### CODE REVIEW

File: builder/vsphere/driver/vm.go [hunk 1]
```diff
@@ -1096,29 +1101,39 @@ func findNetwork(network string, host string, d *VCenterDriver) (object.NetworkR
...
-			// handle error, for example:
-			log.Printf("Error parsing DeviceId: %v\n", err)
-			continue
+			return types.VirtualPCIPassthrough{}, fmt.Errorf("error parsing DeviceId %s: %v", device.DeviceId, err)
 		}
+		if deviceId > math.MaxInt32 || deviceId < math.MinInt32 {
+			return types.VirtualPCIPassthrough{}, fmt.Errorf("DeviceId %s exceeds int32 range", device.DeviceId)
+		}
+
 		vendorId, err := strconv.ParseUint(device.VendorId, 16, 32)
```
Comment (lbajolet-hashicorp) [inline builder/vsphere/driver/vm.go:1115]: Is there a reason why this is a `ParseUint` and not a `ParseInt`? Or the other way around, I would think that `deviceId` cannot be negative, can it?

File: builder/vsphere/driver/vm.go [hunk 2]
```diff
@@ -1096,29 +1101,39 @@ func findNetwork(network string, host string, d *VCenterDriver) (object.NetworkR
...
 	allowedDevices := make([]types.VirtualPCIPassthroughAllowedDevice, len(devices))
 	for i, device := range devices {
 		deviceId, err := strconv.ParseInt(device.DeviceId, 16, 32)
 		if err != nil {
-			// handle error, for example:
-			log.Printf("Error parsing DeviceId: %v\n", err)
-			continue
+			return types.VirtualPCIPassthrough{}, fmt.Errorf("error parsing DeviceId %s: %v", device.DeviceId, err)
 		}
+		if deviceId > math.MaxInt32 || deviceId < math.MinInt32 {
```
Comment (lbajolet-hashicorp) [inline builder/vsphere/driver/vm.go:1111]: Given the documentation of `ParseInt`, I would expect it to error if the returned value cannot fit on 32 bits, so this check seems unnecessary; I understand this silences what was reported, but IMHO this is basically guaranteed not to happen if we handle the error returned by `ParseInt` above

Comment (lbajolet-hashicorp) [reply builder/vsphere/driver/vm.go:1111]: Note: can confirm, this errors in case it is out-of-bound: https://play.golang.com/p/nn3VDOewJM4

**Review by lbajolet-hashicorp — COMMENTED** [review]
Thanks for the PR @tenthirtyam,

At first glance it would seem that the clamping is unnecessary as `ParseInt` already errors if the value can't fit on the provided `bitSize`, that said I understand this is meant to silence the errors reported by CodeQL, but I wonder if there's some other way we can document that those checks don't need to exist, and that they shouldn't be reported.

That said, those checks don't fundamentally harm the code so I'm not completely against merging this, but I think we would benefit from investigating how we can silence this check as it is being too cautious I think.

File: builder/vsphere/driver/vm.go (hunk 2 above)
Comment (tenthirtyam) [reply builder/vsphere/driver/vm.go]: Updated in https://github.com/hashicorp/packer-plugin-vsphere/commit/4acaf070794ee50fed70c457f06a704cd6c98122. Given that `strconv.ParseInt` will indeed return an error if the value cannot fit within the specified bit size, the additional range check is redundant. We can rely on the error handling of `strconv.ParseInt` to catch out-of-bound values.

**Review by lbajolet-hashicorp — APPROVED** [review]
LGTM! Thanks for the reroll @tenthirtyam.

Only thing I can think of left is maybe using `%s` for interpolating errors in the error string, as you did a PR not too long ago to harmonise that, it may be good to continue using `%s` for those use cases.

That said it's very nitpicky so feel free to disregard if you think it's find as-is

Pre-approving to not block later

**Pull request merged**